File: Src/java/elm-fhir/src/main/java/org/cqframework/cql/elm/requirements/ElmExpressionRequirement.java
package org.cqframework.cql.elm.requirements;

import org.hl7.elm.r1.*;

public class ElmExpressionRequirement extends ElmRequirement {
    public ElmExpressionRequirement(VersionedIdentifier libraryIdentifier, Expression expression) {
        super(libraryIdentifier, expression);
    }

    public Expression getExpression() {
        return (Expression)this.element;
    }

    public Expression getElement() {
        return getExpression();
    }

    public ElmExpressionRequirement combine(ElmExpressionRequirement requirement) {
        return this;
    }

    public boolean isLiteral() {
        return this.element instanceof Literal;
    }

    public boolean isTerminologyReference() {
        return this.element instanceof ValueSetRef
                || this.element instanceof CodeSystemRef
                || this.element instanceof ConceptRef
                || this.element instanceof CodeRef;
    }

    public boolean isParameterReference() {
        return this.element instanceof ParameterRef;
    }
}

File: Src/java/elm-fhir/src/main/java/org/cqframework/cql/elm/requirements/ElmExpressionRequirement.kt
package org.cqframework.cql.elm.requirements

import org.hl7.elm.r1.CodeRef
import org.hl7.elm.r1.CodeSystemRef
import org.hl7.elm.r1.ConceptRef
import org.hl7.elm.r1.Expression
import org.hl7.elm.r1.Literal
import org.hl7.elm.r1.ParameterRef
import org.hl7.elm.r1.ValueSetRef
import org.hl7.elm.r1.VersionedIdentifier

open class ElmExpressionRequirement(
    libraryIdentifier: VersionedIdentifier,
    expression: Expression,
) : ElmRequirement(libraryIdentifier, expression) {
    open val expression: Expression
        get() = this.element as Expression

    fun getElement(): Expression {
        return this.expression
    }

    open fun combine(requirement: ElmRequirement?): ElmExpressionRequirement? {
        return this
    }

    open val isLiteral: Boolean
        get() = this.element is Literal

    val isTerminologyReference: Boolean
        get() =
            this.element is ValueSetRef ||
                this.element is CodeSystemRef ||
                this.element is ConceptRef ||
                this.element is CodeRef

    val isParameterReference: Boolean
        get() = this.element is ParameterRef
}
